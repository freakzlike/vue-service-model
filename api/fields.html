<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Fields API | vue-service-model</title>
    <meta name="generator" content="VuePress 1.6.0">
    <link rel="icon" href="/vue-service-model/logo.png">
    <meta name="description" content="Vue.js library for handling REST service requests with caching, aggregation and model definitions">
    <link rel="preload" href="/vue-service-model/assets/css/0.styles.8f06bef0.css" as="style"><link rel="preload" href="/vue-service-model/assets/js/app.8c53950d.js" as="script"><link rel="preload" href="/vue-service-model/assets/js/3.fa9100c0.js" as="script"><link rel="preload" href="/vue-service-model/assets/js/10.b2be555c.js" as="script"><link rel="prefetch" href="/vue-service-model/assets/js/11.50b7caab.js"><link rel="prefetch" href="/vue-service-model/assets/js/12.cf67a250.js"><link rel="prefetch" href="/vue-service-model/assets/js/13.c5391ac2.js"><link rel="prefetch" href="/vue-service-model/assets/js/14.11efe9a3.js"><link rel="prefetch" href="/vue-service-model/assets/js/15.933633f3.js"><link rel="prefetch" href="/vue-service-model/assets/js/16.881caf4f.js"><link rel="prefetch" href="/vue-service-model/assets/js/17.6adf22a7.js"><link rel="prefetch" href="/vue-service-model/assets/js/18.ec72945e.js"><link rel="prefetch" href="/vue-service-model/assets/js/19.55f43912.js"><link rel="prefetch" href="/vue-service-model/assets/js/20.59a2fd09.js"><link rel="prefetch" href="/vue-service-model/assets/js/21.815f43d2.js"><link rel="prefetch" href="/vue-service-model/assets/js/22.11519fc4.js"><link rel="prefetch" href="/vue-service-model/assets/js/23.a1b820b6.js"><link rel="prefetch" href="/vue-service-model/assets/js/24.71a5da24.js"><link rel="prefetch" href="/vue-service-model/assets/js/4.8dad5aaa.js"><link rel="prefetch" href="/vue-service-model/assets/js/5.57995c62.js"><link rel="prefetch" href="/vue-service-model/assets/js/6.41c34acf.js"><link rel="prefetch" href="/vue-service-model/assets/js/7.ca6e4184.js"><link rel="prefetch" href="/vue-service-model/assets/js/8.ea47ba2e.js"><link rel="prefetch" href="/vue-service-model/assets/js/9.e3f94221.js"><link rel="prefetch" href="/vue-service-model/assets/js/vendors~docsearch.e63fde8a.js">
    <link rel="stylesheet" href="/vue-service-model/assets/css/0.styles.8f06bef0.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/vue-service-model/" class="home-link router-link-active"><img src="/vue-service-model/logo.png" alt="vue-service-model" class="logo"> <span class="site-name can-hide">vue-service-model</span></a> <div class="links"><form id="search-form" role="search" class="algolia-search-wrapper search-box"><input id="algolia-search-input" class="search-query"></form> <nav class="nav-links can-hide"><div class="nav-item"><a href="/vue-service-model/guide/" class="nav-link">
  Guide
</a></div><div class="nav-item"><a href="/vue-service-model/api/" class="nav-link router-link-active">
  API Reference
</a></div><div class="nav-item"><a href="https://github.com/freakzlike/vue-service-model/releases" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Release Notes
  <svg xmlns="http://www.w3.org/2000/svg" aria-labelledby="outbound-link-title" role="img" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><title id="outbound-link-title">(opens new window)</title> <path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></div> <a href="https://github.com/freakzlike/vue-service-model" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <svg xmlns="http://www.w3.org/2000/svg" aria-labelledby="outbound-link-title" role="img" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><title id="outbound-link-title">(opens new window)</title> <path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="/vue-service-model/guide/" class="nav-link">
  Guide
</a></div><div class="nav-item"><a href="/vue-service-model/api/" class="nav-link router-link-active">
  API Reference
</a></div><div class="nav-item"><a href="https://github.com/freakzlike/vue-service-model/releases" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Release Notes
  <svg xmlns="http://www.w3.org/2000/svg" aria-labelledby="outbound-link-title" role="img" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><title id="outbound-link-title">(opens new window)</title> <path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></div> <a href="https://github.com/freakzlike/vue-service-model" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <svg xmlns="http://www.w3.org/2000/svg" aria-labelledby="outbound-link-title" role="img" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><title id="outbound-link-title">(opens new window)</title> <path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></nav>  <ul class="sidebar-links"><li><a href="/vue-service-model/" aria-current="page" class="sidebar-link">Home</a></li><li><section class="sidebar-group depth-0"><a href="/vue-service-model/guide/" class="sidebar-heading clickable"><span>Guide</span> <!----></a> <ul class="sidebar-links sidebar-group-items"><li><a href="/vue-service-model/guide/getting-started.html" class="sidebar-link">Getting started</a></li><li><a href="/vue-service-model/guide/installation.html" class="sidebar-link">Installation</a></li><li><a href="/vue-service-model/guide/models.html" class="sidebar-link">Models</a></li><li><a href="/vue-service-model/guide/fields.html" class="sidebar-link">Fields</a></li><li><a href="/vue-service-model/guide/field-types.html" class="sidebar-link">Field types</a></li><li><a href="/vue-service-model/guide/service-model.html" class="sidebar-link">ServiceModel</a></li><li><a href="/vue-service-model/guide/model-manager.html" class="sidebar-link">ModelManager</a></li><li><a href="/vue-service-model/guide/components.html" class="sidebar-link">Components</a></li><li><a href="/vue-service-model/guide/configuration.html" class="sidebar-link">Configuration</a></li></ul></section></li><li><section class="sidebar-group depth-0"><p class="sidebar-heading open"><span>API</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/vue-service-model/api/base-model.html" class="sidebar-link">BaseModel API</a></li><li><a href="/vue-service-model/api/service-model.html" class="sidebar-link">ServiceModel API</a></li><li><a href="/vue-service-model/api/fields.html" aria-current="page" class="active sidebar-link">Fields API</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/vue-service-model/api/fields.html#renderablefield" class="sidebar-link">RenderableField</a></li><li class="sidebar-sub-header"><a href="/vue-service-model/api/fields.html#fielddef" class="sidebar-link">FieldDef</a></li><li class="sidebar-sub-header"><a href="/vue-service-model/api/fields.html#fieldbind" class="sidebar-link">FieldBind</a></li><li class="sidebar-sub-header"><a href="/vue-service-model/api/fields.html#inputprops" class="sidebar-link">InputProps</a></li><li class="sidebar-sub-header"><a href="/vue-service-model/api/fields.html#inputrenderdata" class="sidebar-link">InputRenderData</a></li></ul></li><li><a href="/vue-service-model/api/model-manager.html" class="sidebar-link">ModelManager API</a></li><li><a href="/vue-service-model/api/configuration.html" class="sidebar-link">Configuration API</a></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="fields-api"><a href="#fields-api" class="header-anchor">#</a> Fields API</h1> <div class="language-typescript extra-class"><pre class="language-typescript"><code><span class="token keyword">class</span> <span class="token class-name">Field</span> <span class="token punctuation">{</span>
  <span class="token comment">// Constructor takes field definition</span>
  <span class="token keyword">constructor</span> <span class="token punctuation">(</span>def<span class="token operator">:</span> FieldDef<span class="token punctuation">,</span> fieldBind<span class="token operator">?</span><span class="token operator">:</span> FieldBind<span class="token punctuation">)</span>

  <span class="token comment">// Clone field instance</span>
  <span class="token keyword">public</span> <span class="token function">clone</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> Field

  <span class="token comment">// Returns field name (Which has been set as key at fieldsDef.</span>
  <span class="token comment">// Will throw FieldNotBoundException in case field has not been bound</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">name</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">string</span>

  <span class="token comment">// Returns either attributeName from FieldDef or field name</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">attributeName</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">string</span>

  <span class="token comment">// Returns field definition</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">definition</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> FieldDef

  <span class="token comment">// Assigned model</span>
  <span class="token comment">// Will throw FieldNotBoundException in case field has not been bound to a model</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">model</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> BaseModel

  <span class="token comment">// Return bound data or data from bound model</span>
  <span class="token comment">// Will throw FieldNotBoundException if field is not bound to data or model </span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">data</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> Dictionary<span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Property mapper for getValue()</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">value</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">any</span>

  <span class="token comment">// Returns async field value from data by calling valueGetter with data of assigned model</span>
  <span class="token keyword">public</span> <span class="token keyword">async</span> <span class="token function">getValue</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Property mapper for setValue()</span>
  <span class="token keyword">public</span> <span class="token keyword">set</span> <span class="token function">value</span> <span class="token punctuation">(</span>value<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span>

  <span class="token comment">// Sets field value to model data by calling valueSetter with data of assigned model</span>
  <span class="token keyword">public</span> <span class="token function">setValue</span> <span class="token punctuation">(</span>value<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token keyword">void</span>

  <span class="token comment">// Parses raw value with valueParser and sets field value by calling valueSetter</span>
  <span class="token keyword">public</span> <span class="token keyword">async</span> <span class="token function">setParseValue</span> <span class="token punctuation">(</span>rawValue<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Returns async field label from field definition</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">label</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">string</span><span class="token operator">&gt;</span>

  <span class="token comment">// Returns async field options with validation and default values depending on field type</span>
  <span class="token class-name"><span class="token keyword">public</span></span> <span class="token keyword">get</span> <span class="token function">options</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>FieldTypeOptions<span class="token operator">&gt;</span>

  <span class="token comment">// Validate field options and set default values depending on field type</span>
  <span class="token class-name"><span class="token keyword">protected</span></span> <span class="token keyword">async</span> <span class="token function">validateOptions</span> <span class="token punctuation">(</span>options<span class="token operator">:</span> FieldTypeOptions<span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>FieldTypeOptions<span class="token operator">&gt;</span>

  <span class="token comment">// Returns boolean whether field is a primary key</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">isPrimaryKey</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">boolean</span>

  <span class="token comment">// Returns boolean whether attribute is an nested data structure</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">isNestedAttribute</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">boolean</span>

  <span class="token comment">// Retrieve value from data structure according to attributeName</span>
  <span class="token comment">// Uses nested syntax from attributeName (e.g. &quot;address.city&quot; -&gt; {address: {city: 'New York'}})</span>
  <span class="token comment">// Will return null if value is not available</span>
  <span class="token keyword">public</span> <span class="token function">valueGetter</span> <span class="token punctuation">(</span>data<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">any</span>

  <span class="token comment">// Set value to data by using attributeName</span>
  <span class="token comment">// Will create nested structure from attributeName (e.g. &quot;address.city&quot; -&gt; {address: {city: 'New York'}})</span>
  <span class="token keyword">public</span> <span class="token function">valueSetter</span> <span class="token punctuation">(</span>value<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">,</span> data<span class="token operator">:</span> Dictionary<span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token keyword">void</span>

  <span class="token comment">// Parse a raw value and return the parsed value with valid data type</span>
  <span class="token class-name"><span class="token keyword">public</span></span> <span class="token keyword">async</span> <span class="token function">valueParser</span> <span class="token punctuation">(</span>rawValue<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Map value from a data structure to another data structure.</span>
  <span class="token comment">// Uses valueGetter and valueSetter</span>
  <span class="token keyword">public</span> <span class="token function">mapFieldValue</span> <span class="token punctuation">(</span>fromData<span class="token operator">:</span> Dictionary<span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span><span class="token punctuation">,</span> toData<span class="token operator">:</span> Dictionary<span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token keyword">void</span>
<span class="token punctuation">}</span>
</code></pre></div><h2 id="renderablefield"><a href="#renderablefield" class="header-anchor">#</a> RenderableField</h2> <div class="language-typescript extra-class"><pre class="language-typescript"><code><span class="token keyword">class</span> <span class="token class-name">RenderableField</span> <span class="token keyword">extends</span> <span class="token class-name">Field</span> <span class="token punctuation">{</span>
  <span class="token comment">// Display component to render when displaying value with &lt;display-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">displayComponent</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>ComponentModule<span class="token operator">&gt;</span>

  <span class="token comment">// Async function to prepare before displayRender gets called</span>
  <span class="token comment">// Can return any data which needs to be resolved for displayRender</span>
  <span class="token comment">// renderProps can be passed from DisplayField component </span>
  <span class="token keyword">public</span> <span class="token keyword">async</span> <span class="token function">prepareDisplayRender</span> <span class="token punctuation">(</span>renderProps<span class="token operator">?</span><span class="token operator">:</span> object <span class="token operator">|</span> <span class="token keyword">null</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Simple Vue render function when using default displayComponent when displaying value with &lt;display-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token function">displayRender</span> <span class="token punctuation">(</span>h<span class="token operator">:</span> CreateElement<span class="token punctuation">,</span> renderData<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> VNode

  <span class="token comment">// Input component to render when showing input for field with &lt;input-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">inputComponent</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>ComponentModule<span class="token operator">&gt;</span>

  <span class="token comment">// Async function to prepare before inputRender gets called</span>
  <span class="token comment">// Can return any data which needs to be resolved for inputRender</span>
  <span class="token comment">// renderProps can be passed from InputField component</span>
  <span class="token keyword">public</span> <span class="token keyword">async</span> <span class="token function">prepareInputRender</span> <span class="token punctuation">(</span>inputProps<span class="token operator">:</span> InputProps<span class="token punctuation">,</span> renderProps<span class="token operator">?</span><span class="token operator">:</span> object <span class="token operator">|</span> <span class="token keyword">null</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>InputRenderData<span class="token operator">&gt;</span>

  <span class="token comment">// Simple Vue render function when using default inputComponent for input of field value with &lt;input-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token function">inputRender</span> <span class="token punctuation">(</span>h<span class="token operator">:</span> CreateElement<span class="token punctuation">,</span> renderData<span class="token operator">:</span> InputRenderData<span class="token punctuation">)</span><span class="token operator">:</span> VNode
<span class="token punctuation">}</span>
</code></pre></div><h2 id="fielddef"><a href="#fielddef" class="header-anchor">#</a> FieldDef</h2> <div class="language-typescript extra-class"><pre class="language-typescript"><code><span class="token keyword">interface</span> <span class="token class-name">FieldDef</span> <span class="token punctuation">{</span>
  <span class="token comment">// String which key should be used to retrieve value from.</span>
  <span class="token comment">// See Attribute name for more information</span>
  <span class="token comment">// Optional: default uses key from fieldsDef</span>
  attributeName<span class="token operator">?</span><span class="token operator">:</span> <span class="token builtin">string</span>

  <span class="token comment">// Label of field. See Field label for more information</span>
  <span class="token comment">// Optional: Can either be a string, function or promise</span>
  label<span class="token operator">?</span><span class="token operator">:</span> LazyValue<span class="token operator">&lt;</span><span class="token builtin">string</span><span class="token operator">&gt;</span>

  <span class="token comment">// Boolean flag whether field is a primary key</span>
  <span class="token comment">// Optional: default is false</span>
  primaryKey<span class="token operator">?</span><span class="token operator">:</span> <span class="token builtin">boolean</span>

  <span class="token comment">// Optional field type specific options.</span>
  <span class="token comment">// Can either be an object, a function or a promise.</span>
  <span class="token comment">// See field Types</span>
  options<span class="token operator">?</span><span class="token operator">:</span> LazyValue<span class="token operator">&lt;</span>FieldTypeOptions<span class="token operator">&gt;</span>
<span class="token punctuation">}</span>
</code></pre></div><h2 id="fieldbind"><a href="#fieldbind" class="header-anchor">#</a> FieldBind</h2> <div class="language-typescript extra-class"><pre class="language-typescript"><code><span class="token keyword">interface</span> <span class="token class-name">FieldBind</span> <span class="token punctuation">{</span>
  <span class="token comment">// Field name which has been set as key at fieldsDef</span>
  name<span class="token operator">?</span><span class="token operator">:</span> <span class="token builtin">string</span>

  <span class="token comment">// Model to bind field to</span>
  <span class="token comment">// Beware of manual bind your field to a model afterwards.</span>
  <span class="token comment">// The field will not be available by model.fields or model.getField()</span>
  model<span class="token operator">?</span><span class="token operator">:</span> BaseModel

  <span class="token comment">// Data to retrieve field value from</span>
  <span class="token comment">// If not set then model.data will be used</span>
  data<span class="token operator">?</span><span class="token operator">:</span> Dictionary<span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Value which should be bound to field</span>
  <span class="token comment">// When using value then model and data will be ignored</span>
  <span class="token comment">// A custom data object will be created to keep the field reactive</span>
  value<span class="token operator">?</span><span class="token operator">:</span> <span class="token builtin">any</span>
<span class="token punctuation">}</span>
</code></pre></div><h2 id="inputprops"><a href="#inputprops" class="header-anchor">#</a> InputProps</h2> <div class="language-typescript extra-class"><pre class="language-typescript"><code><span class="token keyword">interface</span> <span class="token class-name">InputProps</span> <span class="token punctuation">{</span>
  <span class="token comment">// Input flag for disabled input</span>
  disabled<span class="token operator">:</span> <span class="token builtin">boolean</span>

  <span class="token comment">// Input flag for readonly input</span>
  <span class="token keyword">readonly</span><span class="token operator">:</span> <span class="token builtin">boolean</span>
<span class="token punctuation">}</span>
</code></pre></div><h2 id="inputrenderdata"><a href="#inputrenderdata" class="header-anchor">#</a> InputRenderData</h2> <div class="language-typescript extra-class"><pre class="language-typescript"><code><span class="token keyword">interface</span> <span class="token class-name">InputRenderData</span> <span class="token punctuation">{</span>
  <span class="token comment">// Resolved value of field</span>
  value<span class="token operator">:</span> <span class="token builtin">any</span>

  <span class="token comment">// Input properties</span>
  inputProps<span class="token operator">:</span> InputProps
<span class="token punctuation">}</span>
</code></pre></div></div> <footer class="page-edit"><div class="edit-link"><a href="https://github.com/freakzlike/vue-service-model/edit/master/docs/api/fields.md" target="_blank" rel="noopener noreferrer">Edit this page on GitHub</a> <svg xmlns="http://www.w3.org/2000/svg" aria-labelledby="outbound-link-title" role="img" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><title id="outbound-link-title">(opens new window)</title> <path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></div> <div class="last-updated"><span class="prefix">Last Updated:</span> <span class="time">9/26/2020, 8:15:34 AM</span></div></footer> <div class="page-nav"><p class="inner"><span class="prev">
      ←
      <a href="/vue-service-model/api/service-model.html" class="prev">
        ServiceModel API
      </a></span> <span class="next"><a href="/vue-service-model/api/model-manager.html">
        ModelManager API
      </a>
      →
    </span></p></div> </main></div><div class="global-ui"></div></div>
    <script src="/vue-service-model/assets/js/app.8c53950d.js" defer></script><script src="/vue-service-model/assets/js/3.fa9100c0.js" defer></script><script src="/vue-service-model/assets/js/10.b2be555c.js" defer></script>
  </body>
</html>
