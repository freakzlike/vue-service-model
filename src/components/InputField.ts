import { VNode, defineComponent, DefineComponent, h, toRaw } from 'vue'
import FieldPropsMixin from './FieldPropsMixin'
import LoadingSlotMixin from './LoadingSlotMixin'

interface ComponentData {
  inputComponent: DefineComponent | null
}

/**
 * Main component to display value of a field
 */
export default defineComponent({
  name: 'InputField',
  inheritAttrs: false,
  mixins: [FieldPropsMixin, LoadingSlotMixin],

  props: {
    disabled: {
      type: Boolean,
      default: false
    },
    readonly: {
      type: Boolean,
      default: false
    }
  },

  data: (): ComponentData => ({
    inputComponent: null
  }),

  // TODO: vue-async-computed
  // asyncComputed: {
  //   async inputComponent (): Promise<DefineComponent | null> {
  //     configHandler.checkWarningUseAsyncComputed()
  //     return this.resolveInputComponent()
  //   }
  // },

  watch: {
    fieldObj () {
      this.setResolveInputComponent()
    }
  },

  created () {
    this.setResolveInputComponent()
  },

  methods: {
    async setResolveInputComponent () {
      // if (!configHandler.useAsyncComputed()) {
      this.inputComponent = await this.resolveInputComponent()
      // }
    },

    async resolveInputComponent () {
      if (this.fieldObj) {
        const componentModule = await this.fieldObj.inputComponent
        return componentModule.default
      } else {
        return null
      }
    }
  },

  render (): VNode {
    if (this.inputComponent && this.fieldObj) {
      return h(toRaw(this.inputComponent), {
        field: this.fieldObj,
        renderProps: this.renderProps,
        disabled: this.disabled,
        readonly: this.readonly
      })
    } else {
      return this.renderLoading()
    }
  }
})
