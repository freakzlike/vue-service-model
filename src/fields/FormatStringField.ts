import { isNull } from '../utils/common'
import { RenderableField } from './RenderableField'

export class FormatStringField extends RenderableField {
  /**
   * Prepare formatted value for display render
   */
  public async prepareDisplayRender (): Promise<any> {
    return this.valueFormatter()
  }

  /**
   * Format the value to a string
   */
  public async valueFormatter (): Promise<string | null> {
    const value = await this.value
    return !isNull(value) ? String(value) : null
  }
}
