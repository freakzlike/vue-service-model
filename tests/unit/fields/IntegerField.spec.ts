import { BaseModel } from '@/models/BaseModel'
import { IntegerField } from '@/fields/IntegerField'
import { FormatStringField } from '@/fields/FormatStringField'
import { mount } from '@vue/test-utils'
import InputField from '@/components/InputField'
import { waitRender } from '../../testUtils'
import { RenderableField } from '@/fields'

describe('fields/IntegerField', () => {
  class TestModel extends BaseModel {
    static fieldsDef = {
      amount: new IntegerField()
    }
  }

  describe('constructor', () => {
    it('should inherit from FormatStringField', async () => {
      const model = new TestModel({})
      const field = model.getField('amount')

      expect(field).toBeInstanceOf(FormatStringField)
    })
  })

  describe('inputRender', () => {
    it('should render correct input field', async () => {
      const model = new TestModel({ amount: 17 })
      const wrapper = mount(InputField, {
        props: {
          field: model.getField('amount') as RenderableField
        }
      })

      await waitRender.InputField()

      expect(wrapper.html()).toMatchSnapshot()

      const inputElement = wrapper.get('input')
      expect(inputElement.element.value).toBe('17')

      model.val.amount = 19

      await waitRender.InputFieldUpdate()

      expect(wrapper.html()).toMatchSnapshot()
      expect(inputElement.element.value).toBe('19')
    })

    it('should set value correct on change', async () => {
      const model = new TestModel({ amount: 17 })
      const wrapper = mount(InputField, {
        props: {
          field: model.getField('amount') as RenderableField
        }
      })

      await waitRender.InputField()

      const inputElement = wrapper.get('input')
      expect(inputElement.element.value).toBe('17')

      inputElement.setValue(19)

      // Wait for value parser
      await wrapper.vm.$nextTick()

      expect(await model.val.amount).toBe(19)

      inputElement.setValue('20')

      // Wait for value parser
      await wrapper.vm.$nextTick()

      expect(await model.val.amount).toBe(20)
    })

    it('should render disabled input field', async () => {
      const model = new TestModel({ amount: 17 })
      const wrapper = mount(InputField, {
        props: {
          field: model.getField('amount') as RenderableField,
          disabled: true
        }
      })

      await waitRender.InputField()

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('should render readonly input field', async () => {
      const model = new TestModel({ amount: 17 })
      const wrapper = mount(InputField, {
        props: {
          field: model.getField('amount') as RenderableField,
          readonly: true
        }
      })

      await waitRender.InputField()

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('valueParser', () => {
    const field = new TestModel({}).getField('amount')

    it('should return integer value', async () => {
      expect(await field.valueParser(7)).toBe(7)
      expect(await field.valueParser('8')).toBe(8)
      expect(await field.valueParser(0)).toBe(0)
      expect(await field.valueParser(9.2)).toBe(9)
    })

    it('should return null', async () => {
      expect(await field.valueParser(null)).toBe(null)
      expect(await field.valueParser(undefined)).toBe(null)
    })
  })
})
