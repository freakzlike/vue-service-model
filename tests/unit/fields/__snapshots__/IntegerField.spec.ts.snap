// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fields/IntegerField inputRender should render correct input field 1`] = `<input type="number">`;

exports[`fields/IntegerField inputRender should render correct input field 2`] = `<input type="number">`;

exports[`fields/IntegerField inputRender should render disabled input field 1`] = `<input type="number" disabled="">`;

exports[`fields/IntegerField inputRender should render readonly input field 1`] = `<input type="number" readonly="">`;
